src/cargo/sources/directory.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::HashMap;
+use std::fmt::{self, Debug, Formatter};
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+use rustc_serialize::hex::ToHex;
+use rustc_serialize::json;
+
+use core::{Package, PackageId, Summary, SourceId, Source, Dependency, Registry};
+use sources::PathSource;
+use util::{CargoResult, human, ChainError, Config, Sha256};
+use util::paths;
+
+pub struct DirectorySource<'cfg> {
+    id: SourceId,
+    root: PathBuf,
+    packages: HashMap<PackageId, (Package, Checksum)>,
+    config: &'cfg Config,
+}
+
+#[derive(RustcDecodable)]
+struct Checksum {
+    package: String,
+    files: HashMap<String, String>,
+}
+
+impl<'cfg> DirectorySource<'cfg> {
+    pub fn new(path: &Path, id: &SourceId, config: &'cfg Config)
+               -> DirectorySource<'cfg> {
+        DirectorySource {
+            id: id.clone(),
+            root: path.to_path_buf(),
+            config: config,
+            packages: HashMap::new(),
+        }
+    }
+}
+
+impl<'cfg> Debug for DirectorySource<'cfg> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "DirectorySource {{ root: {:?} }}", self.root)
+    }
+}
+
+impl<'cfg> Registry for DirectorySource<'cfg> {
+    fn query(&mut self, dep: &Dependency) -> CargoResult<Vec<Summary>> {
+        let packages = self.packages.values().map(|p| &p.0);
+        let matches = packages.filter(|pkg| dep.matches(pkg.summary()));
+        let summaries = matches.map(|pkg| pkg.summary().clone());
+        Ok(summaries.collect())
+    }
+
+    fn supports_checksums(&self) -> bool {
+        true
+    }
+}
+
+impl<'cfg> Source for DirectorySource<'cfg> {
+    fn update(&mut self) -> CargoResult<()> {
+        self.packages.clear();
+        let entries = try!(self.root.read_dir().chain_error(|| {
+            human(format!("failed to read root of directory source: {}",
+                          self.root.display()))
+        }));
+
+        for entry in entries {
```

Comment:
Shouldn't this loop be skipping over the `index` directory, which is not a package?